---
title: "RobustCNV"
author: "Samuel S Hunter (shunter@gmail.com)"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{RobustCNV}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  \usepackage[utf8](inputenc)
---

RobustCNV is a strategy for identifying copy number variants using sparse, noisy targeted capture mapping depth coverage information. This strategy attempts to remove sytematic bias from the mapping depth information prior to segmentation and calling. Systematic bias is removed by fitting a tumor sample against a panel of normal samples using a robust regression. The $log2(observed.tumor.coverage/predicted.tumor.coverage)$ ratio is then calculated, and a GC bias correction is done using local polynomial regression (loess) with the per-interval GC percentages. The resulting values often have less systemtic bias.

Following normalization, the values are segmented using DNAcopy. Segment gains and losses can then be called using a cutoff based on the variability of the segments.

**Note that this is very much a "toy dataset". Real targeted capture coverage data is not nearly as well behaved.**

## Usage example
### Creating a simulated dataset

The following code defines a function to simulate copy number data. In order to simulate systematic bias, we will use a sine function. Random noise is added, and a base coverage of 20x is assumed. 
```{r,}
require(RobustCNV)
sinf = function(x, sd=0, CNV=F, begin=0, end=0, magnitude=1, gc){
  #add systematic and random noise
  m= 3*(sin(x/100)) + rnorm(n=1, mean=0, sd=sd)
  m = m + 20
  #Simulate gc bias:
  m = m * (1 + gc^3 - gc^4)
  #add CNV
  if(CNV & x >= begin & x <= end){
    m = m + abs(m) * magnitude
  }
  
  return(m)
}
```

### Now create test data including: 

* Interval definition table to simulate a real sequencing experiment:
    + Chrom: chromosome 
    + Gene: a gene identifier (should be the same for all intervals comprising a gene)
    + pGC: gc content of the interval
    + Start and End positions
    + Index: a unique numeric index
* A panel of normals containing 5 normals
* Three tumor samples with different types of copy number variants.
* A fourth highly aneuploid tumor and expected allele frequencies calculated as af = alt/(ref + alt).

```{r,}

intervals = data.frame(Chrom=rep(1:23, each=250), 
                      Gene=rep(paste('gene', 1:1150, sep=''), each=5),
                      pGC=sample(x=seq(0,1, .001), size=5750, replace = T),
                      Start=rep(1:250, 23) * 120,
                      End=rep(1:250, 23) * 120 + 70,
                      Index=1:(250*23))

normals = data.frame(n1 = sapply(seq(1,5750, 1), function(x){sinf(x, sd=1, gc=intervals$pGC[x])}), 
                     n2 = sapply(seq(1,5750, 1), function(x){sinf(x, sd=1, gc=intervals$pGC[x])}), 
                     n3 = sapply(seq(1,5750, 1), function(x){sinf(x, sd=1, gc=intervals$pGC[x])}),
                     n4 = sapply(seq(1,5750, 1), function(x){sinf(x, sd=1, gc=intervals$pGC[x])}),
                     n5 = sapply(seq(1,5750, 1), function(x){sinf(x, sd=1, gc=intervals$pGC[x])}))

high_gain_tumor = unlist(sapply(seq(1,5750, 1), function(x)
                                  {sinf(x, sd=1, CNV=T, begin=400, end=500, magnitude=1, gc=intervals$pGC[x])}))
low_gain_tumor = unlist(sapply(seq(1, 5750, 1), function(x)
                                 {sinf(x, sd=1, CNV=T, begin=400, end=500, magnitude=.5,gc=intervals$pGC[x])}))
low_loss_tumor = unlist(sapply(seq(1, 5750, 1), function(x)
                                 {sinf(x, sd=1, CNV=T, begin=50, end=150, magnitude=-.5, gc=intervals$pGC[x])}))
aneuploid_tumor = c(unlist(sapply(seq(1,1000, 1), function(x)
                                  {sinf(x, sd=1, CNV=T, begin=0, end=500, magnitude=.5, gc=intervals$pGC[x])})),
                    unlist(sapply(seq(1001,2000, 1), function(x)
                                  {sinf(x, sd=1, CNV=T, begin=1001, end=1500, magnitude=.5, gc=intervals$pGC[x])})),
                    unlist(sapply(seq(2001,3000, 1), function(x)
                                  {sinf(x, sd=1, CNV=T, begin=2001, end=2500, magnitude=-.5, gc=intervals$pGC[x])})),
                    unlist(sapply(seq(3001,4000, 1), function(x)
                                  {sinf(x, sd=1, CNV=T, begin=3001, end=3500, magnitude=.5, gc=intervals$pGC[x])})),
                    unlist(sapply(seq(4001,5000, 1), function(x)
                                  {sinf(x, sd=1, CNV=T, begin=4001, end=4500, magnitude=.5, gc=intervals$pGC[x])})),
                    unlist(sapply(seq(5001,5750, 1), function(x)
                                  {sinf(x, sd=1, CNV=T, begin=5001, end=5500, magnitude=.5, gc=intervals$pGC[x])}))
                    )

aneuploid_tumor_alleles = c(sample(c(2/3, 1/3), replace = T, size = 500), rep(.5, 500),
                            sample(c(2/3, 1/3), replace = T, size = 500), rep(.5, 500),
                            rep(1, 500), rep(.5, 500),
                            sample(c(2/3, 1/3), replace = T, size = 500), rep(.5, 500),
                            sample(c(2/3, 1/3), replace = T, size = 500), rep(.5, 500),
                            sample(c(2/3, 1/3), replace = T, size = 500), rep(.5, 250)
                            )

aneuploid_tumor_alleles = data.frame(Index=intervals$Index, af=aneuploid_tumor_alleles)
```

### Show that GC bias has been added to the data:
```{r, fig.show='hold'}
plotGCBias("GC bias plot", intervals$pGC, normals[,1])
plotGCBias("GC bias plot", intervals$pGC, high_gain_tumor)
```

### Examin clustering of samples.
Although not very informative in this example, the systematic variation in real datasets can cause tumor and normal samples to cluster together. A nicely interspersed clustering pattern indicates that this systematic bias has been captured in the normal samples. Segregation in the clusters between tumor and normal samples can indicate that systematic bias in the tumor samples is not well represented in the normals. Distances based on Spearman rank correlation are used to reduce the impact of large CNV events on clustering.
```{r, fig.width=6, fig.height=5}
plotSampleClusters(rawNormalCoverage=normals, 
                   rawTumorCoverage=cbind(high_gain_tumor, low_gain_tumor, low_loss_tumor, aneuploid_tumor))
```

### Normalization
Normalize the $tumor$ samples with robustNorm(). Note that interval lengths (passed in as EVs - Explanatory Variables) were not modeled in this example and are all set to the same length. If other explanatory variables are available these might also be passed in.
```{r,}
high_gain_tumor.normalized <- robustNorm(high_gain_tumor, normals, pGC=intervals$pGC, EVs=rep(100, 5750))
low_gain_tumor.normalized <- robustNorm(low_gain_tumor, normals, pGC=intervals$pGC, EVs=rep(100, 5750))
low_loss_tumor.normalized <- robustNorm(low_loss_tumor, normals, pGC=intervals$pGC, EVs=rep(100, 5750))
aneuploid_tumor.normalized <- robustNorm(aneuploid_tumor, normals, pGC=intervals$pGC, EVs=rep(100, 5750))
```

Plots of the pre and post-normalized data show that much of the systematic bias has been removed.
```{r, fig.show='hold'}
plot(high_gain_tumor, main="High gain raw data", pch='.', xlab='index', ylab='')
plot(high_gain_tumor.normalized, main="High gain normalized", pch='.', xlab='index', ylab='')
abline(h=0)
plotGCBias("GC bias plot", intervals$pGC, high_gain_tumor.normalized)
```

```{r, fig.show='hold'}
plot(low_gain_tumor, main="Low gain raw data", pch='.', xlab='index', ylab='')
plot(low_gain_tumor.normalized, main="Low gain normalized", pch='.', xlab='index', ylab='')
abline(h=0)
plotGCBias("GC bias plot", intervals$pGC, low_gain_tumor.normalized)
```

```{r, fig.show='hold'}
plot(low_loss_tumor, main="Low loss raw data", pch='.', xlab='index', ylab='')
plot(low_loss_tumor.normalized, main="Low loss normalized", pch='.', xlab='index', ylab='')
abline(h=0)
plotGCBias("GC bias plot", intervals$pGC, low_gain_tumor.normalized)
```

```{r, fig.show='hold'}
plot(aneuploid_tumor, main="Aneuploid raw data", pch='.', xlab='index', ylab='')
plot(aneuploid_tumor.normalized, main="Aneuploid normalized", pch='.', xlab='index', ylab='')
abline(h=0)
plotGCBias("GC bias plot", intervals$pGC, low_gain_tumor.normalized)
```

Note that the highly aneuploid example is not very well centered. In many cases this can be fixed by using the allele fractions as an indicator of the diploid segments.

```{r, fig.show='hold'}
aneuploid_tumor.recentered = allelicRecentering(aneuploid_tumor.normalized, intervals, aneuploid_tumor_alleles)

plot(aneuploid_tumor, main="Aneuploid raw data", pch='.', xlab='index', ylab='')
plot(aneuploid_tumor.recentered, main="Aneuploid normalized and recentered", pch='.', xlab='index', ylab='')
abline(h=0)
plotGCBias("GC bias plot", intervals$pGC, low_gain_tumor.normalized)
```


### Segmentation

Segment normalized data and call gains and losses.
```{r,}

high_gain_tumor.segmented <- segment(high_gain_tumor.normalized, 
                                     orderedChrom=as.ordered(intervals$Chrom), maploc=intervals$End)
low_gain_tumor.segmented <- segment(low_gain_tumor.normalized, 
                                    orderedChrom=as.ordered(intervals$Chrom), maploc=intervals$End)
low_loss_tumor.segmented <- segment(low_loss_tumor.normalized, 
                                    orderedChrom=as.ordered(intervals$Chrom), maploc=intervals$End)
aneuploid_tumor.segmented <- segment(aneuploid_tumor.recentered, 
                                    orderedChrom=as.ordered(intervals$Chrom), maploc=intervals$End)

```

### Calling gains and losses

```{r, }
high_gain_tumor.called <- callSegments(high_gain_tumor.segmented, 
                                       high_gain_tumor.normalized, tuning=0.8)
low_gain_tumor.called <- callSegments(low_gain_tumor.segmented, 
                                      low_gain_tumor.normalized, tuning=0.8)
low_loss_tumor.called <- callSegments(low_loss_tumor.segmented, 
                                      low_loss_tumor.normalized, tuning=0.8)
aneuploid_tumor.called <- callSegments(aneuploid_tumor.segmented, 
                                      aneuploid_tumor.recentered, tuning=0.8)

```

### Assigning segment calls to intervals for easier reporting
```{r, fig.width=7, fig.height=5}
high_gain_tumor.called.intervals <- assignIntervalCalls(intervals, high_gain_tumor.called, high_gain_tumor.normalized)
low_gain_tumor.called.intervals <- assignIntervalCalls(intervals, low_gain_tumor.called, low_gain_tumor.normalized)
low_loss_tumor.called.intervals <- assignIntervalCalls(intervals, low_loss_tumor.called, low_loss_tumor.normalized)
aneuploid_tumor.called.intervals <- assignIntervalCalls(intervals, aneuploid_tumor.called, aneuploid_tumor.recentered)

plotGenomeIntervalCalls(perIntervalCalls=high_gain_tumor.called.intervals, main='High gain tumor')
plotGenomeIntervalCalls(perIntervalCalls=low_gain_tumor.called.intervals, main='Low gain tumor')
plotGenomeIntervalCalls(perIntervalCalls=low_loss_tumor.called.intervals, main='Low loss tumor')
plotGenomeIntervalCalls(perIntervalCalls=aneuploid_tumor.called.intervals, main='Aneuploid tumor')

```

### 
